content/en/blog/2023/testing-otel-demo/index.md

```diff
@@ -0,0 +1,399 @@
+  - selector: span[name="oteldemo.ShippingService/ShipOrder" rpc.system="grpc"
+			rpc.method="ShipOrder" rpc.service="oteldemo.ShippingService"]
+    name: "The product was shipped"
+    assertions:
+    - attr:rpc.grpc.status_code = 0
+    - attr:selected_spans.count >= 1
+  - selector: span[name="oteldemo.CartService/EmptyCart" rpc.system="grpc"
+			rpc.method="EmptyCart" rpc.service="oteldemo.CartService"]
+    name: "The cart was emptied"
+    assertions:
+    - attr:rpc.grpc.status_code = 0
+    - attr:selected_spans.count >= 1
+```
+
+Finally, when running these tests we have the following report, showing each
+test file run in the transaction and also with the "Checkout” step described
+above:
+
+![results from the trace-based test](./trace-based-test-result.png)
+
+## Running the tests and evaluating the OpenTelemetry Demo
+
+With the completed test suite, run it by executing `make run-tracetesting` in
+the demo. This will evaluate all services in the OpenTelemetry Demo.
+
+During the development of the tests, we noticed some differences in the test
+results. For example, some minor fixes were made to the Cypress tests, and some
+behaviors were observed in the back-end APIs that can be tested and investigated
+at a later time. You can find the details in
+[this pull request](https://github.com/open-telemetry/opentelemetry-demo/pull/950)
+and
+[this discussion](https://github.com/open-telemetry/opentelemetry-demo/pull/905#discussion_r1207101535).
+
+One curious case was the behavior of the
+[EmailService](https://github.com/open-telemetry/opentelemetry-demo/tree/main/src/emailservice).
+When building tests for the first time and calling it directly with the payload
+provided by AVA tests, traces were generated for the service indicating success,
+but with an HTTP `500` error, as can be seen in Jaeger.
+
+![single-email-formatted.png](./single-email-formatted.png)
+
+However, running it as part of the checkout process it executed as expected, as
+shown in this Jaeger screenshot:
+
+![email-under-checkout-formatted.png](./email-under-checkout-formatted.png)
+
+What could have happened? By looking deeper at the telemetry and the code, we
+found out that due to the nature of handling email templates on the Email
+service, written in Ruby using the `snake_case` standard, instead of sending the
+order details as `pascalCase` in `JSON`:
+
+```json
+{
+  "email": "google@example.com",
+  "order": {
+    "orderId": "505",
+    "shippingCost": {
+      "currencyCode": "USD",
+    },
+	  // ...
+}
+```
+
+We should pass them as `snake_case`, which the Checkout service does correctly:
+
+```json
+{
+  "email": "google@example.com",
+  "order": {
+    "order_id": "505",
+    "shipping_cost": {
+      "currency_code": "USD",
+    },
+    // ...
+}
+```
+
+By doing that, we have a successful call to the service, and it evaluates
+correctly as seen here:
+
+![email-success-formatted.png](./email-success-formatted.png)
```

Comment:
Is there any reason why tracetest is used in this last screenshot and not jaeger or a console output as before? Also, the picture is really hard to parse on a small screen

Reply:
The reasoning is to show the test specs and that the test is passing. I can try re-taking the screenshot and make it bigger.